sql/opt: TestAggregateIgnoresDuplicates failed

sql/opt.TestAggregateIgnoresDuplicates [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/15069672?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/15069672?buildTab=artifacts#/) on release-24.1 @ [03021e93be411b533ec53a2c84623df92782d165](https://github.com/cockroachdb/cockroach/commits/03021e93be411b533ec53a2c84623df92782d165):

Fatal error:

```
panic: test timed out after 57s
running tests:
	TestAggregateIgnoresDuplicates (57s)

```
Stack: 

```
goroutine 46144 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x30c
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestAggregateIgnoresDuplicates
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/72d18d9aee7439da5cb8b913bb76e21f/logTestAggregateIgnoresDuplicates731033170
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:157: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAggregateIgnoresDuplicates.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-38347

--- a/pkg/sql/opt/operator_agg_test.go
+++ b/pkg/sql/opt/operator_agg_test.go
@@ -13,6 +13,7 @@ package opt_test
 import (
 	"context"
 	"fmt"
+	"sort"
 	"testing"
 
 	"github.com/cockroachdb/cockroach/pkg/base"
@@ -110,6 +111,11 @@ func TestAggregateIgnoresDuplicates(t *testing.T) {
 		// st_extent
 		{op: opt.STExtentOp, typ: types.Geometry},
 	}
+	// Re-order test cases so that all cases that use the same type are
+	// contiguous (to avoid recreating the table).
+	sort.Slice(testCases, func(i, j int) bool {
+		return testCases[i].typ.SQLString() < testCases[j].typ.SQLString()
+	})
 
 	// Ensure that a test case exists for each operator that
 	// AggregateIgnoresDuplicates returns true for.
@@ -143,16 +149,22 @@ func TestAggregateIgnoresDuplicates(t *testing.T) {
 		numIters = 10
 	)
 	rng, _ := randutil.NewTestRand()
+	var prevTyp *types.T
 	for _, tc := range testCases {
 		sqlOp, ok := opt.AggregateOpReverseMap[tc.op]
 		if !ok {
 			t.Fatalf("%s is not an aggregate function", tc.op.String())
 		}
 
-		for i := 0; i < numDatums; i++ {
-			// Create the results table.
+		// Create the results table if the previous one used a different type.
+		if prevTyp == nil || !prevTyp.Identical(tc.typ) {
 			tDB.Exec(t, "DROP TABLE IF EXISTS results")
 			tDB.Exec(t, fmt.Sprintf("CREATE TABLE results (r %s)", tc.typ.SQLString()))
+			prevTyp = tc.typ
+		}
+		for i := 0; i < numDatums; i++ {
+			// Clear the table.
+			tDB.Exec(t, "DELETE FROM results WHERE true")
 
 			// Generate a random datum.
 			datum := randgen.RandDatum(rng, tc.typ, false /* nullOk */)

